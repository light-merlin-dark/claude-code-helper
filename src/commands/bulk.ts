/**
 * Bulk operation commands for efficient management across many projects
 * Supports pattern matching and --all flag for project selection
 */

import { loadClaudeConfig, saveClaudeConfig, ClaudeConfig } from '../core/config';
import { logger } from '../utils/logger';
import { promptConfirm } from '../utils/prompt';
import { matchProjects, projectMatchesPatterns } from '../utils/patterns';
import { createBackup } from '../services/backup';
import { DANGEROUS_PATTERNS } from '../services/analyzer';
import { ProgressBar, withProgress } from '../utils/progress';

export interface BulkAddPermissionOptions {
  permission: string;
  projects?: string | string[];
  all?: boolean;
  testMode?: boolean;
  dryRun?: boolean;
}

export interface BulkRemovePermissionOptions {
  permission?: string;
  dangerous?: boolean;
  projects?: string | string[];
  all?: boolean;
  testMode?: boolean;
  dryRun?: boolean;
}

export interface BulkAddToolOptions {
  tool: string;
  projects?: string | string[];
  all?: boolean;
  testMode?: boolean;
  dryRun?: boolean;
}

export interface BulkRemoveToolOptions {
  tool: string;
  projects?: string | string[];
  all?: boolean;
  testMode?: boolean;
  dryRun?: boolean;
}

export interface BulkOperationResult {
  projectsModified: number;
  itemsAdded?: number;
  itemsRemoved?: number;
  backupPath?: string;
}

/**
 * Add permission to multiple projects
 */
export async function bulkAddPermission(options: BulkAddPermissionOptions): Promise<BulkOperationResult> {
  const { permission, projects, all, testMode = false, dryRun = false } = options;
  
  if (!permission) {
    throw new Error('Permission is required');
  }
  
  // Create backup first
  let backupPath: string | undefined;
  if (!dryRun && !testMode) {
    backupPath = await createBackup('pre-bulk-add-perm');
  }
  
  const config = await loadClaudeConfig(testMode) as ClaudeConfig;
  
  // Get projects to modify
  const projectsToModify = getProjectsToModify(config, projects, all);
  
  if (projectsToModify.length === 0) {
    throw new Error('No projects matched the criteria');
  }
  
  // Show what will be modified
  if (!dryRun) {
    console.log(`\nAdding permission "${permission}" to ${projectsToModify.length} projects:`);
    projectsToModify.slice(0, 5).forEach(([name]) => console.log(`  - ${name}`));
    if (projectsToModify.length > 5) {
      console.log(`  ... and ${projectsToModify.length - 5} more`);
    }
    
    const confirmed = await promptConfirm('Continue?', true);
    if (!confirmed) {
      throw new Error('Operation cancelled');
    }
  }
  
  let itemsAdded = 0;
  let projectsModified = 0;
  
  // Use progress bar for large operations
  const useProgress = projectsToModify.length > 10 && !dryRun;
  const progress = useProgress ? new ProgressBar(projectsToModify.length, 'Adding permission') : null;
  
  projectsToModify.forEach(([projectName, project], index) => {
    if (!project.allowedCommands) {
      project.allowedCommands = [];
    }
    
    // Check if permission already exists
    if (!project.allowedCommands.includes(permission)) {
      if (!dryRun) {
        project.allowedCommands.push(permission);
      }
      itemsAdded++;
      projectsModified++;
      
      if (!useProgress) {
        logger.info(`${dryRun ? '[DRY RUN] Would add' : 'Added'} "${permission}" to ${projectName}`);
      }
    }
    
    if (progress) {
      progress.increment();
    }
  });
  
  if (!dryRun && itemsAdded > 0) {
    await saveClaudeConfig(config, testMode);
  }
  
  return {
    projectsModified,
    itemsAdded,
    backupPath
  };
}

/**
 * Remove permission from multiple projects
 */
export async function bulkRemovePermission(options: BulkRemovePermissionOptions): Promise<BulkOperationResult> {
  const { permission, dangerous, projects, all, testMode = false, dryRun = false } = options;
  
  if (!permission && !dangerous) {
    throw new Error('Either permission or --dangerous flag is required');
  }
  
  // Create backup first
  let backupPath: string | undefined;
  if (!dryRun && !testMode) {
    backupPath = await createBackup('pre-bulk-remove-perm');
  }
  
  const config = await loadClaudeConfig(testMode) as ClaudeConfig;
  
  // Get projects to modify
  const projectsToModify = getProjectsToModify(config, projects, all);
  
  if (projectsToModify.length === 0) {
    throw new Error('No projects matched the criteria');
  }
  
  let itemsRemoved = 0;
  let projectsModified = 0;
  
  if (dangerous) {
    // Remove all dangerous permissions
    console.log('\nScanning for dangerous permissions...');
    
    const dangerousFound = new Map<string, string[]>();
    
    projectsToModify.forEach(([projectName, project]) => {
      if (project.allowedCommands) {
        const dangerousPerms = project.allowedCommands.filter((cmd: string) =>
          isDangerousCommand(cmd)
        );
        
        if (dangerousPerms.length > 0) {
          dangerousFound.set(projectName, dangerousPerms);
        }
      }
    });
    
    if (dangerousFound.size === 0) {
      console.log('No dangerous permissions found');
      return { projectsModified: 0, itemsRemoved: 0 };
    }
    
    // Show what will be removed
    console.log(`\nFound dangerous permissions in ${dangerousFound.size} projects:`);
    let count = 0;
    dangerousFound.forEach((perms, projectName) => {
      if (count < 5) {
        console.log(`  ${projectName}:`);
        perms.forEach(perm => console.log(`    - ${perm}`));
      }
      count++;
    });
    if (dangerousFound.size > 5) {
      console.log(`  ... and ${dangerousFound.size - 5} more projects`);
    }
    
    if (!dryRun) {
      const confirmed = await promptConfirm('Remove all dangerous permissions?', true);
      if (!confirmed) {
        throw new Error('Operation cancelled');
      }
    }
    
    // Remove dangerous permissions
    dangerousFound.forEach((perms, projectName) => {
      const project = config.projects![projectName];
      if (!dryRun && project.allowedCommands) {
        project.allowedCommands = project.allowedCommands.filter((cmd: string) =>
          !isDangerousCommand(cmd)
        );
      }
      itemsRemoved += perms.length;
      projectsModified++;
    });
  } else if (permission) {
    // Remove specific permission
    console.log(`\nRemoving permission "${permission}" from matching projects...`);
    
    projectsToModify.forEach(([projectName, project]) => {
      if (project.allowedCommands && project.allowedCommands.includes(permission)) {
        if (!dryRun) {
          project.allowedCommands = project.allowedCommands.filter((cmd: string) => cmd !== permission);
        }
        itemsRemoved++;
        projectsModified++;
        
        logger.info(`${dryRun ? '[DRY RUN] Would remove' : 'Removed'} "${permission}" from ${projectName}`);
      }
    });
  }
  
  if (!dryRun && itemsRemoved > 0) {
    await saveClaudeConfig(config, testMode);
  }
  
  return {
    projectsModified,
    itemsRemoved,
    backupPath
  };
}

/**
 * Add MCP tool to multiple projects
 */
export async function bulkAddTool(options: BulkAddToolOptions): Promise<BulkOperationResult> {
  const { tool, projects, all, testMode = false, dryRun = false } = options;
  
  if (!tool) {
    throw new Error('Tool name is required');
  }
  
  // Create backup first
  let backupPath: string | undefined;
  if (!dryRun && !testMode) {
    backupPath = await createBackup('pre-bulk-add-tool');
  }
  
  const config = await loadClaudeConfig(testMode) as ClaudeConfig;
  
  // Get projects to modify
  const projectsToModify = getProjectsToModify(config, projects, all);
  
  if (projectsToModify.length === 0) {
    throw new Error('No projects matched the criteria');
  }
  
  // Parse MCP tool format (e.g., "github" or "mcp__github")
  const mcpName = tool.startsWith('mcp__') ? tool.split('__')[1] : tool;
  
  console.log(`\nAdding MCP tool "${mcpName}" to ${projectsToModify.length} projects`);
  
  if (!dryRun) {
    const confirmed = await promptConfirm('Continue?', true);
    if (!confirmed) {
      throw new Error('Operation cancelled');
    }
  }
  
  let itemsAdded = 0;
  let projectsModified = 0;
  
  projectsToModify.forEach(([projectName, project]) => {
    if (!project.mcpServers) {
      project.mcpServers = {};
    }
    
    if (!project.mcpServers[mcpName]) {
      if (!dryRun) {
        // Add with empty config - user can configure later
        project.mcpServers[mcpName] = {};
      }
      itemsAdded++;
      projectsModified++;
      
      logger.info(`${dryRun ? '[DRY RUN] Would add' : 'Added'} MCP "${mcpName}" to ${projectName}`);
    }
  });
  
  if (!dryRun && itemsAdded > 0) {
    await saveClaudeConfig(config, testMode);
  }
  
  return {
    projectsModified,
    itemsAdded,
    backupPath
  };
}

/**
 * Remove MCP tool from multiple projects
 */
export async function bulkRemoveTool(options: BulkRemoveToolOptions): Promise<BulkOperationResult> {
  const { tool, projects, all, testMode = false, dryRun = false } = options;
  
  if (!tool) {
    throw new Error('Tool name is required');
  }
  
  // Create backup first
  let backupPath: string | undefined;
  if (!dryRun && !testMode) {
    backupPath = await createBackup('pre-bulk-remove-tool');
  }
  
  const config = await loadClaudeConfig(testMode) as ClaudeConfig;
  
  // Get projects to modify
  const projectsToModify = getProjectsToModify(config, projects, all);
  
  if (projectsToModify.length === 0) {
    throw new Error('No projects matched the criteria');
  }
  
  // Parse MCP tool format
  const mcpName = tool.startsWith('mcp__') ? tool.split('__')[1] : tool;
  
  console.log(`\nRemoving MCP tool "${mcpName}" from matching projects...`);
  
  let itemsRemoved = 0;
  let projectsModified = 0;
  
  projectsToModify.forEach(([projectName, project]) => {
    if (project.mcpServers && project.mcpServers[mcpName]) {
      if (!dryRun) {
        delete project.mcpServers[mcpName];
      }
      itemsRemoved++;
      projectsModified++;
      
      logger.info(`${dryRun ? '[DRY RUN] Would remove' : 'Removed'} MCP "${mcpName}" from ${projectName}`);
    }
  });
  
  if (!dryRun && itemsRemoved > 0) {
    await saveClaudeConfig(config, testMode);
  }
  
  return {
    projectsModified,
    itemsRemoved,
    backupPath
  };
}

/**
 * Get projects to modify based on patterns or --all flag
 */
function getProjectsToModify(
  config: ClaudeConfig,
  patterns?: string | string[],
  all?: boolean
): Array<[string, any]> {
  if (all) {
    return Object.entries(config.projects || {});
  }
  
  if (patterns) {
    return matchProjects(patterns, config);
  }
  
  // No patterns and no --all flag
  throw new Error('Please specify project patterns or use --all flag');
}

/**
 * Check if a command is dangerous
 */
function isDangerousCommand(command: string): boolean {
  return DANGEROUS_PATTERNS.some(pattern => {
    const regexPattern = pattern
      .replace(/\*/g, '.*')
      .replace(/\?/g, '.')
      .replace(/\[/g, '\\[')
      .replace(/\]/g, '\\]');
    
    const regex = new RegExp(`^${regexPattern}$`);
    return regex.test(command);
  });
}