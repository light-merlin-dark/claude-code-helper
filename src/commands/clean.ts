/**
 * Clean commands for removing bloat and dangerous permissions
 */

import fs from 'fs';
import { createGzip } from 'zlib';
import { pipeline } from 'stream/promises';
import { loadClaudeConfig, saveClaudeConfig, ClaudeConfig } from '../core/config';
import { getBackupsDir } from '../core/paths';
import { logger } from '../utils/logger';
import { DANGEROUS_PATTERNS } from '../services/analyzer';
import { matchProjects } from '../utils/patterns';

export interface CleanHistoryOptions {
  projects?: string[];
  testMode?: boolean;
  dryRun?: boolean;
}

export interface CleanHistoryResult {
  pastesRemoved: number;
  sizeReduction: number;
  projectsModified: number;
  backupPath?: string;
}

export interface CleanDangerousOptions {
  testMode?: boolean;
  dryRun?: boolean;
}

export interface CleanDangerousResult {
  permissionsRemoved: number;
  projectsModified: number;
  backupPath?: string;
}

/**
 * Clean large pastes from conversation history
 */
export async function cleanHistory(options: CleanHistoryOptions = {}): Promise<CleanHistoryResult> {
  const { projects = [], testMode = false, dryRun = false } = options;
  
  // Create backup first
  let backupPath: string | undefined;
  if (!dryRun && !testMode) {
    backupPath = await createBackup('pre-clean');
  }
  
  const config = await loadClaudeConfig(testMode) as ClaudeConfig;
  const originalSize = Buffer.byteLength(JSON.stringify(config), 'utf8');
  
  let pastesRemoved = 0;
  let projectsModified = 0;
  
  // Get projects to clean
  const projectsToClean = projects.length > 0 
    ? matchProjects(projects, config)
    : Object.entries(config.projects || {});
  
  for (const [projectName, project] of projectsToClean) {
    if (project) {
      const largePastes = findLargePastes(project.history || []);
      
      if (largePastes.length > 0) {
        if (!dryRun) {
          project.history = removePastes(project.history || [], largePastes);
        }
        pastesRemoved += largePastes.length;
        projectsModified++;
        
        logger.info(`${dryRun ? '[DRY RUN] Would clean' : 'Cleaned'} ${largePastes.length} pastes from ${projectName}`);
      }
    }
  }
  
  if (!dryRun && pastesRemoved > 0) {
    await saveClaudeConfig(config, testMode);
  }
  
  const newSize = Buffer.byteLength(JSON.stringify(config), 'utf8');
  const sizeReduction = originalSize - newSize;
  
  return {
    pastesRemoved,
    sizeReduction,
    projectsModified,
    backupPath
  };
}

/**
 * Remove dangerous permissions from all projects
 */
export async function cleanDangerous(options: CleanDangerousOptions = {}): Promise<CleanDangerousResult> {
  const { testMode = false, dryRun = false } = options;
  
  // Create backup first
  let backupPath: string | undefined;
  if (!dryRun && !testMode) {
    backupPath = await createBackup('pre-clean-dangerous');
  }
  
  const config = await loadClaudeConfig(testMode) as ClaudeConfig;
  
  let permissionsRemoved = 0;
  let projectsModified = 0;
  
  Object.entries(config.projects || {}).forEach(([projectName, project]) => {
    if (project && project.allowedCommands) {
      const originalCount = project.allowedCommands.length;
      
      if (!dryRun) {
        project.allowedCommands = project.allowedCommands.filter(command => {
          return !isDangerousCommand(command);
        });
      } else {
        // Count dangerous commands for dry run
        const dangerousCount = project.allowedCommands.filter(command => 
          isDangerousCommand(command)
        ).length;
        
        if (dangerousCount > 0) {
          permissionsRemoved += dangerousCount;
          projectsModified++;
        }
      }
      
      if (!dryRun && project.allowedCommands.length < originalCount) {
        const removed = originalCount - project.allowedCommands.length;
        permissionsRemoved += removed;
        projectsModified++;
        
        logger.info(`Removed ${removed} dangerous permissions from ${projectName}`);
      }
    }
  });
  
  if (!dryRun && permissionsRemoved > 0) {
    await saveClaudeConfig(config, testMode);
  }
  
  return {
    permissionsRemoved,
    projectsModified,
    backupPath
  };
}

interface PasteInfo {
  entryIndex: number;
  pasteId: string;
  lines: number;
  size: number;
}

function findLargePastes(history: any[]): PasteInfo[] {
  const largePastes: PasteInfo[] = [];
  
  history.forEach((entry, entryIndex) => {
    if (entry.pastedContents) {
      Object.entries(entry.pastedContents).forEach(([pasteId, paste]: [string, any]) => {
        const content = paste.content || '';
        const lines = content.split('\n').length;
        const size = Buffer.byteLength(content, 'utf8');
        
        if (lines > 100) { // Threshold for large pastes
          largePastes.push({
            entryIndex,
            pasteId,
            lines,
            size
          });
        }
      });
    }
  });
  
  return largePastes;
}

function removePastes(history: any[], pastesToRemove: PasteInfo[]): any[] {
  // Group pastes by entry index
  const pastesByEntry = new Map<number, Set<string>>();
  pastesToRemove.forEach(paste => {
    if (!pastesByEntry.has(paste.entryIndex)) {
      pastesByEntry.set(paste.entryIndex, new Set());
    }
    pastesByEntry.get(paste.entryIndex)!.add(paste.pasteId);
  });
  
  // Create a new history with pastes removed
  return history.map((entry, index) => {
    if (!pastesByEntry.has(index) || !entry.pastedContents) {
      return entry;
    }
    
    const pastesToRemoveFromEntry = pastesByEntry.get(index)!;
    const newEntry = { ...entry };
    
    // Remove specific pastes
    const newPastedContents: any = {};
    Object.entries(entry.pastedContents).forEach(([id, content]) => {
      if (!pastesToRemoveFromEntry.has(id)) {
        newPastedContents[id] = content;
      }
    });
    
    // Update display text if needed
    if (Object.keys(newPastedContents).length === 0) {
      // All pastes removed, clean up display text
      delete newEntry.pastedContents;
      if (newEntry.display) {
        newEntry.display = newEntry.display.replace(/\[Pasted text[^\]]+\]/g, '[Pasted content removed]');
      }
    } else {
      newEntry.pastedContents = newPastedContents;
    }
    
    return newEntry;
  });
}

function isDangerousCommand(command: string): boolean {
  return DANGEROUS_PATTERNS.some(pattern => {
    const regexPattern = pattern
      .replace(/\*/g, '.*')
      .replace(/\?/g, '.')
      .replace(/\[/g, '\\[')
      .replace(/\]/g, '\\]');
    
    const regex = new RegExp(`^${regexPattern}$`);
    return regex.test(command);
  });
}

async function createBackup(prefix: string): Promise<string> {
  const timestamp = new Date().toISOString().replace(/:/g, '-').split('.')[0];
  const backupDir = getBackupsDir();
  const backupPath = `${backupDir}/${prefix}-${timestamp}.json.gz`;
  
  // Ensure backup directory exists
  if (!fs.existsSync(backupDir)) {
    fs.mkdirSync(backupDir, { recursive: true });
  }
  
  // Read config and compress
  const config = await loadClaudeConfig() as ClaudeConfig;
  const configStr = JSON.stringify(config, null, 2);
  
  // Create gzip stream and write compressed data
  const gzip = createGzip({ level: 9 });
  const destination = fs.createWriteStream(backupPath);
  
  // Create a promise to handle the compression
  await new Promise<void>((resolve, reject) => {
    destination.on('finish', resolve);
    destination.on('error', reject);
    gzip.on('error', reject);
    
    // Pipe gzip output to file
    gzip.pipe(destination);
    
    // Write config to gzip and end the stream
    gzip.write(configStr);
    gzip.end();
  });
  
  logger.info(`Backup created: ${backupPath}`);
  return backupPath;
}