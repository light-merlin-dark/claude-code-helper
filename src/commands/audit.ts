/**
 * Audit command - comprehensive configuration analysis
 * Provides security analysis, bloat detection, and optimization recommendations
 */

import { Analyzer } from '../services/analyzer';
import { loadClaudeConfig } from '../core/config';
import { logger } from '../utils/logger';
import { promptConfirm } from '../utils/prompt';
import { backupConfig } from './config/backup';
import { cleanHistory, cleanDangerous } from './clean';
import { withProgress } from '../utils/progress';

export interface AuditOptions {
  fix?: boolean;
  stats?: boolean;
  showSecrets?: boolean;
  testMode?: boolean;
}

export async function audit(options: AuditOptions = {}): Promise<string> {
  const { fix = false, stats = false, showSecrets = false, testMode = false } = options;

  try {
    // Read Claude configuration
    const config = await withProgress(
      'Loading configuration',
      () => loadClaudeConfig(testMode)
    );
    
    // Run analysis
    const analyzer = new Analyzer();
    const report = await withProgress(
      'Analyzing configuration',
      () => analyzer.analyzeConfig(config)
    );
    
    if (fix) {
      // Interactive fix mode
      return await interactiveFix(report, testMode);
    }
    
    if (stats) {
      // Quick stats mode for AI agents
      return formatQuickStats(config, report);
    }
    
    if (showSecrets) {
      // Show detailed secret information
      return formatSecretDetails(report);
    }
    
    // Return formatted report
    return analyzer.formatReport(report);
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown error';
    logger.error(`Audit failed: ${message}`);
    throw error;
  }
}

async function interactiveFix(report: any, testMode: boolean): Promise<string> {
  const results: string[] = [];
  
  console.log('\n🔧 Interactive Fix Mode\n');
  
  // First, create a backup
  const shouldBackup = await promptConfirm(
    'Create a backup before making changes?',
    true
  );
  
  if (shouldBackup) {
    console.log('Creating backup...');
    await backupConfig('pre-audit-fix', testMode);
    results.push('✓ Backup created');
  }
  
  // Fix dangerous permissions
  if (report.security.length > 0) {
    console.log(`\n⚠️  Found ${report.security.length} dangerous permissions`);
    const shouldClean = await promptConfirm(
      'Remove all dangerous permissions?',
      true
    );
    
    if (shouldClean) {
      const cleanResult = await cleanDangerous({ testMode });
      results.push(`✓ Removed dangerous permissions from ${cleanResult.projectsModified} projects`);
    }
  }
  
  // Fix config bloat
  if (report.bloat.totalPastes > 0) {
    const sizeInMB = (report.bloat.totalSize / 1024 / 1024).toFixed(1);
    console.log(`\n📦 Found ${report.bloat.totalPastes} large pastes (${sizeInMB} MB)`);
    
    // Get projects with bloat
    const bloatedProjects = Array.from(report.bloat.projectSummaries.keys()) as string[];
    
    const shouldCleanHistory = await promptConfirm(
      `Clean history for ${bloatedProjects.length} projects?`,
      true
    );
    
    if (shouldCleanHistory) {
      const cleanResult = await cleanHistory({
        projects: bloatedProjects,
        testMode
      });
      
      const reductionMB = (cleanResult.sizeReduction / 1024 / 1024).toFixed(1);
      results.push(`✓ Cleaned ${cleanResult.pastesRemoved} pastes, saved ${reductionMB} MB`);
    }
  }
  
  // Summary
  console.log('\n✅ Audit fixes completed:');
  results.forEach(result => console.log(`  ${result}`));
  
  return results.join('\n');
}

function formatQuickStats(config: any, report: any): string {
  const configSize = Buffer.byteLength(JSON.stringify(config), 'utf8');
  const sizeMB = (configSize / 1024 / 1024).toFixed(1);
  const projectCount = Object.keys(config.projects || {}).length;
  
  // Count issues
  const securityIssues = report.security.length;
  const secretsFound = report.secrets.totalCount;
  const highConfidenceSecrets = report.secrets.highConfidenceCount;
  const bloatIssues = report.bloat.totalPastes;
  const totalIssues = securityIssues + (secretsFound > 0 ? 1 : 0) + (bloatIssues > 10 ? 1 : 0) + (configSize > 50 * 1024 * 1024 ? 1 : 0);
  
  // Performance assessment
  let performanceStatus = 'Good';
  if (configSize > 100 * 1024 * 1024) {
    performanceStatus = 'Poor';
  } else if (configSize > 50 * 1024 * 1024) {
    performanceStatus = 'Degraded';
  }
  
  const stats = [
    `Config: ${sizeMB}MB`,
    `Projects: ${projectCount}`,
    `Issues: ${totalIssues}`,
    `Performance: ${performanceStatus}`
  ];
  
  // Priority: secrets first (most critical)
  if (secretsFound > 0) {
    if (highConfidenceSecrets > 0) {
      stats.push(`🚨 Secrets: ${highConfidenceSecrets} high-confidence`);
    } else {
      stats.push(`🔍 Secrets: ${secretsFound} potential`);
    }
  }
  
  if (securityIssues > 0) {
    stats.push(`Security: ${securityIssues} dangerous permissions`);
  }
  
  if (bloatIssues > 10) {
    stats.push(`Bloat: ${bloatIssues} large pastes`);
  }
  
  // Prioritized recommendations
  let recommendation = 'Config is healthy';
  if (highConfidenceSecrets > 0) {
    recommendation = '🚨 URGENT: Run "cch --clean-config --mask-secrets" to secure secrets';
  } else if (secretsFound > 0) {
    recommendation = 'Run "cch --audit --show-secrets" to review potential secrets';
  } else if (totalIssues > 0) {
    recommendation = 'Run "cch --clean-config" to optimize';
  }
  
  return `${stats.join(', ')} - ${recommendation}`;
}

function formatSecretDetails(report: any): string {
  const lines: string[] = [];
  
  if (report.secrets.totalCount === 0) {
    return '✅ No secrets detected in configuration';
  }
  
  lines.push('🔐 DETECTED SECRETS REPORT');
  lines.push('===============================');
  lines.push('');
  lines.push(`Total secrets found: ${report.secrets.totalCount}`);
  lines.push(`High confidence: ${report.secrets.highConfidenceCount}`);
  lines.push(`Medium/Low confidence: ${report.secrets.totalCount - report.secrets.highConfidenceCount}`);
  lines.push('');
  
  // Group secrets by confidence level
  const highConfidence = report.secrets.secrets.filter((s: any) => s.confidence === 'high');
  const mediumConfidence = report.secrets.secrets.filter((s: any) => s.confidence === 'medium');
  const lowConfidence = report.secrets.secrets.filter((s: any) => s.confidence === 'low');
  
  if (highConfidence.length > 0) {
    lines.push('🚨 HIGH CONFIDENCE SECRETS (Immediate Action Required):');
    lines.push('───────────────────────────────────────────────────────');
    highConfidence.forEach((secret: any, index: number) => {
      lines.push(`${index + 1}. ${secret.type}`);
      lines.push(`   Location: ${secret.location}`);
      lines.push(`   Value: ${secret.maskedValue}`);
      lines.push(`   Context: ${secret.context.substring(0, 100)}...`);
      lines.push('');
    });
  }
  
  if (mediumConfidence.length > 0) {
    lines.push('⚠️  MEDIUM CONFIDENCE SECRETS (Review Recommended):');
    lines.push('──────────────────────────────────────────────────');
    mediumConfidence.slice(0, 5).forEach((secret: any, index: number) => {
      lines.push(`${index + 1}. ${secret.type} in ${secret.location}`);
      lines.push(`   Value: ${secret.maskedValue}`);
      lines.push('');
    });
    if (mediumConfidence.length > 5) {
      lines.push(`   ... and ${mediumConfidence.length - 5} more medium confidence secrets`);
      lines.push('');
    }
  }
  
  if (lowConfidence.length > 0) {
    lines.push('ℹ️  LOW CONFIDENCE SECRETS (May be false positives):');
    lines.push('────────────────────────────────────────────────────');
    const grouped = lowConfidence.reduce((acc: any, secret: any) => {
      acc[secret.type] = (acc[secret.type] || 0) + 1;
      return acc;
    }, {});
    
    Object.entries(grouped).forEach(([type, count]) => {
      lines.push(`   ${type}: ${count} instances`);
    });
    lines.push('');
  }
  
  lines.push('RECOMMENDED ACTIONS:');
  lines.push('──────────────────────');
  if (highConfidence.length > 0) {
    lines.push('• URGENT: Run "cch --clean-config --mask-secrets" to mask high-confidence secrets');
  }
  if (report.secrets.totalCount > 0) {
    lines.push('• Create backup: "cch --backup-config -n pre-secret-cleanup"');
    lines.push('• Review and manually verify any false positives');
  }
  lines.push('• Consider using environment variables for sensitive configuration');
  
  return lines.join('\n');
}