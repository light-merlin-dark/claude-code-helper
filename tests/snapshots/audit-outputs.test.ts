import { describe, test, expect, beforeEach, afterEach } from 'bun:test';
import { audit } from '../../src/commands/audit';
import { 
  setupTestConfig, 
  cleanupTestConfig, 
  TEST_CONFIGS 
} from '../test-data-utils';

describe('Audit Output Snapshots', () => {
  let testWorkspace: string;

  afterEach(async () => {
    if (testWorkspace) {
      await cleanupTestConfig(testWorkspace);
    }
  });

  test('clean config audit output', async () => {
    testWorkspace = await setupTestConfig(TEST_CONFIGS.CLEAN, 'snapshot-clean');
    
    const result = await audit({
      configPath: `${testWorkspace}/.claude.json`,
      testMode: true
    });

    expect(result).toMatchSnapshot();
  });

  test('dangerous config audit output', async () => {
    testWorkspace = await setupTestConfig(TEST_CONFIGS.DANGEROUS, 'snapshot-dangerous');
    
    const result = await audit({
      configPath: `${testWorkspace}/.claude.json`,
      testMode: true
    });

    expect(result).toMatchSnapshot();
  });

  test('bloated config audit output', async () => {
    testWorkspace = await setupTestConfig(TEST_CONFIGS.BLOATED, 'snapshot-bloated');
    
    const result = await audit({
      configPath: `${testWorkspace}/.claude.json`,
      testMode: true
    });

    expect(result).toMatchSnapshot();
  });
});