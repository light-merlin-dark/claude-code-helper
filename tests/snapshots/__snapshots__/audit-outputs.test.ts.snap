// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit Output Snapshots clean config audit output 1`] = `
"===========================================
Claude Code Configuration Audit Report
===========================================

OVERVIEW:
- Total projects: 4
- Total config size: 172 B
- MCP tools installed: 0
- Total permissions: 0

PROJECT TREE:
├── work/ (2 projects)
│   ├── api
│   └── frontend
├── personal/ (1 projects)
│   └── blog
└── test-api

RECOMMENDATIONS:
1. Backup before changes: cch backup

Run 'cch audit --fix' to interactively resolve issues."
`;

exports[`Audit Output Snapshots dangerous config audit output 1`] = `
"===========================================
Claude Code Configuration Audit Report
===========================================

OVERVIEW:
- Total projects: 4
- Total config size: 172 B
- MCP tools installed: 0
- Total permissions: 0

PROJECT TREE:
├── work/ (2 projects)
│   ├── api
│   └── frontend
├── personal/ (1 projects)
│   └── blog
└── test-api

RECOMMENDATIONS:
1. Backup before changes: cch backup

Run 'cch audit --fix' to interactively resolve issues."
`;

exports[`Audit Output Snapshots bloated config audit output 1`] = `
"===========================================
Claude Code Configuration Audit Report
===========================================

OVERVIEW:
- Total projects: 4
- Total config size: 172 B
- MCP tools installed: 0
- Total permissions: 0

PROJECT TREE:
├── work/ (2 projects)
│   ├── api
│   └── frontend
├── personal/ (1 projects)
│   └── blog
└── test-api

RECOMMENDATIONS:
1. Backup before changes: cch backup

Run 'cch audit --fix' to interactively resolve issues."
`;
